monai/transforms/intensity/array.py

```diff
@@ -1292,3 +1293,288 @@ def _apply_mask(self, k: np.ndarray) -> np.ndarray:
+            images with 3D spatial dimensions, the user can provide (C, X, Y, Z)
+            to fix which channel C is affected, or (X, Y, Z) to place the same
+            spike in all channels. For 2D cases, the user can provide (C, X, Y)
+            or (X, Y).
+        k_itensity (float, Sequence(float)): value for the log-intensity of the
+            `k`-space version of the image. This value should be tested. The
+            default values are the 2.5 the mean of the log intensity for each
+            channel.
+        as_tensor_output (bool): if ``True`` return torch.Tensor, else return np.array.
+            Default: ``True``.
+
+    Example:
+        When working with 4D data, ``KSpaceSpikeNoise(loc = ((3,60,64,32), (64,60,32)), k_intensity = (13,14))``
+        will place a spike at `[3, 60, 64, 32]` with `log-intensity = 13`, and
+        four spikes located at `[: , 64, 60, 32]` with `log-intensity = 14`.
+    """
+
+    def __init__(
+        self,
+        loc: Union[Tuple, Sequence[Tuple]],
+        k_intensity: Union[float, Sequence[float]] = None,
+        as_tensor_output: bool = True,
+    ):
+
+        self.loc = ensure_tuple(loc)
+        self.as_tensor_output = as_tensor_output
+        self.k_intensity = k_intensity
+        self._device = torch.device("cpu")
+
+        # assert one-to-one relationship between factors and locations
+        if isinstance(k_intensity, Sequence):
+            if not isinstance(loc[0], Sequence):
+                raise AssertionError(
+                    "If a sequence is passed to k_intensity, then a sequence of locations must be passed to loc"
+                )
+            elif len(self.k_intensity) != len(loc):
+                raise AssertionError("There must be one intensity_factor value for each tuple of indices in loc.")
+        if isinstance(self.loc[0], Sequence) and k_intensity is not None:
+            if not isinstance(self.k_intensity, Sequence):
+                raise AssertionError("There must be one intensity_factor value for each tuple of indices in loc.")
+
+    def __call__(self, img: Union[np.ndarray, torch.Tensor]) -> Union[torch.Tensor, np.ndarray]:
+        """
+        Args:
+            img (np.array or torch.tensor): image with dimensions (C, H, W) or (C, H, W, D)
+        """
+
+        n_dims = len(img.shape[1:])
+
+        if len(img.shape) < 3:
+            raise AssertionError("Image needs a channel direction.")
+        if isinstance(self.loc[0], int) and len(img.shape) == 4 and len(self.loc) == 2:
+            raise AssertionError("Input images of dimension 4 need location tuple to be length 3 or 4")
+        if isinstance(self.loc[0], Sequence) and len(img.shape) == 4 and min(map(lambda x: len(x), self.loc)) == 2:
+            raise AssertionError("Input images of dimension 4 need location tuple to be length 3 or 4")
+
+        # convert to ndarray to work with np.fft
+        if isinstance(img, torch.Tensor):
+            self._device = img.device
```

Comment:
just a minor comment here -- the instance variables such as `self._device` and `self.k_intensity` are not thread-safe, could you please make them local, something like `_device = img.device` or inherit `ThreadUnsafe` like https://github.com/Project-MONAI/MONAI/blob/dev/monai/transforms/spatial/array.py#L392. it's fine for the `RandomizableTransform`, as `ThreadUnsafe` is one of the base classes of it.

Reply:
Ok, will do!